architecture/2025-02-application-autoscaling.md

```diff
@@ -0,0 +1,239 @@
+    ports: {
+    }
+    runtimes: {
+      kubernetes: {
+        autoscaling: {
+          minReplicas: 1
+          maxReplicas: 10
+          cpuUtilization: 50
+          memoryLimit: 50      
+        }
+      }
+    }
+  }
+}
+```
+
+Eg : ACA HTTP based autoscaling config
+
+```bicep
+resource container 'Applications.Containers@2023-10-01-preview' = {
+  name: 'myapp'
+  properties: {
+    image: 'myregistry.azurecr.io/myapp:latest'
+    ports: {
+    }
+    runtimes: {
+      serverless-platforms: {
+        autoscaling: {
+          minReplicas: 1
+          maxReplicas: 10
+          http-concurrency: 50  
+        }
+      }
+    }
+  }
+}
+```
+
+Pros: 
+1. Easy to understand and configure.
+1. Autoscale config is run-time specific and doesn't have to be standardized across runtimes as different platforms have different autoscaling mechanisms. 
+1. Enterprises pick platforms based on their application workloads and the underlying platform provides default autoscaling policies based on the workloads. Radius provides a clear abstraction to leverage default autoscaling mechanisms from the underlying platform.
+
+Cons:
+1. More configuration to manage across different runtimes.
+2. The developer needs to know the underlying platform and the autoscaling policies required for the platform.
+
+#### Option 2: Adding it as part extensions
+
+`extensions` have been the way to provide punch through capabilities which do not change the behaviors of the resource. The following is an example of adding autoscaling as part of the extensions. 
+
+```bicep
+resource container 'Applications.Containers@2023-10-01-preview' = {
+  name: 'myapp'
+  properties: {
+    image: 'myregistry.azurecr.io/myapp:latest'
+    extensions: [
+      {
+        kind: 'autoscaling'
+        minReplicas: 1
+        maxReplicas: 10
+        cpuUtilization: 50
+        memoryLimit: 50
+        httpConcurrency: 50
+        }
+      }
+    ]
+  }
+}
+```
+
+Pros: consistent with existing manual scaling config
+
+Cons: not straightforward or intuitive as user doesn't know which autoscale config is required for the platform ? needs a platform specific discriminator
+May be, this is not a developer problem and should be handled by the platform operator. 
+
+#### Option 3: Adding it as top level common property
+
+```bicep
+resource container 'Applications.Containers@2023-10-01-preview' = {
+  name: 'myapp'
+  properties: {
+    image: 'myregistry.azurecr.io/myapp:latest'
+    autoscaling: {
+      minReplicas: 1
+      maxReplicas: 10
+      hpa: {
+        cpuUtilization: 50
+        memoryLimit: 50
+      }
+      http: {
+        httpConcurrency: 50
+      }  
+    }
+  }
+}
+``` 
+
+Cons:same as above, not straightforward or intuitive as user doesn't know which autoscale config is required for the platform ? May be, this is not a developer problem and should be handled by the platform operator.
+
+### Configuring Autoscaling in Radius Environment
```

Comment:
Would autoscaling policies specified in the container override the autoscaling policies set for the environment?